### PR TITLE
JAMES-3693 Rate limiting for Apache James on top of Redis sentinel - introduce docker-compose sample

### PR DESCRIPTION
**Pull request by vttranlina**
Why require commit ? [JAMES-3693 Fixup RedisConfiguration - parsing correct `redisURL` when multiple redis sentinel endpoint](https://github.com/apache/james-project/commit/344a80157059fb18d2830da3c0eaaea6f11ca53c)

The current James configuration uses:
```
config.setListDelimiterHandler(new DefaultListDelimiterHandler(','))
```
Then when give `redisURL` is `redis-sentinel://secret1@redis-sentinel-1:26379,redis-sentinel-2:26379,redis-sentinel-3:26379?sentinelMasterId=mymaster` (multiple sentinel endpoint)
the James configuration parsing is incorrect.

### CODE REVIEW

**Comment by vttranlina** [comment]
Note: my manual test:

- :white_check_mark: I tested the case of running docker-compose up, then shut down the redis-master node. The failover of the sentinel and replica nodes was completed successfully. During the failover process, the log from James:

```
2024-07-24T19:23:26.149474096Z 19:23:26.149 [INFO ] o.a.j.w.m.LoggingRequestFilter - WebAdmin request received
2024-07-24T19:23:27.940731528Z 19:23:27.940 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:27.941470385Z 19:23:27.941 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:28.137933072Z 19:23:28.137 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:28.138330538Z 19:23:28.138 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:28.334451052Z 19:23:28.334 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:28.334743380Z 19:23:28.334 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:28.539857676Z 19:23:28.539 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:28.540234393Z 19:23:28.540 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:28.743990148Z 19:23:28.743 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:28.744325638Z 19:23:28.744 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:28.932333544Z 19:23:28.932 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:28.932631133Z 19:23:28.932 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:29.150144421Z 19:23:29.150 [WARN ] o.a.j.w.r.HealthCheckRoutes - HealthCheck is unstable for Redis : Can not connect to Redis.
2024-07-24T19:23:29.150510257Z 19:23:29.150 [INFO ] o.a.j.w.m.LoggingResponseFilter - WebAdmin response received
2024-07-24T19:23:34.551245574Z 19:23:34.550 [WARN ] i.l.c.p.ConnectionWatchdog - Cannot reconnect to [172.19.0.2/<unresolved>:6379]: java.nio.channels.ClosedChannelException
2024-07-24T19:23:34.551261975Z java.nio.channels.ClosedChannelException: null
2024-07-24T19:23:34.551265471Z 	at io.netty.channel.epoll.AbstractEpollChannel.doClose(AbstractEpollChannel.java:159)
2024-07-24T19:23:34.551268547Z 	at io.netty.channel.epoll.AbstractEpollStreamChannel.doClose(AbstractEpollStreamChannel.java:673)
2024-07-24T19:23:34.551271402Z 	at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:752)
2024-07-24T19:23:34.551274338Z 	at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:729)
2024-07-24T19:23:34.551277283Z 	at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:619)
2024-07-24T19:23:34.551280279Z 	at io.netty.channel.DefaultChannelPipeline$HeadContext.close(DefaultChannelPipeline.java:1349)
2024-07-24T19:23:34.551282974Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:755)
2024-07-24T19:23:34.551285830Z 	at io.netty.channel.AbstractChannelHandlerContext.access$1200(AbstractChannelHandlerContext.java:61)
2024-07-24T19:23:34.551288785Z 	at io.netty.channel.AbstractChannelHandlerContext$11.run(AbstractChannelHandlerContext.java:738)
2024-07-24T19:23:34.551291230Z 	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
2024-07-24T19:23:34.551293734Z 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
2024-07-24T19:23:34.551296279Z 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
2024-07-24T19:23:34.551299014Z 	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:405)
2024-07-24T19:23:34.551301659Z 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
2024-07-24T19:23:34.551309604Z 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2024-07-24T19:23:34.551314573Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2024-07-24T19:23:34.551324562Z 	at java.base/java.lang.Thread.run(Unknown Source)
2024-07-24T19:23:56.332209814Z 19:23:56.332 [INFO ] o.a.j.w.m.LoggingRequestFilter - WebAdmin request received
```
The health check degrades for a few seconds before recovering to a healthy state
=> confirming it works as expected. 

Regarding the rate limiter integration test, I have already pushed the commit.[org.apache.james.rate.limiter.RedisRateLimiterWithSentinelTest#rateLimitShouldWorkNormallyAfterFailoverComplete](https://github.com/apache/james-project/pull/2358/commits/b130942fd4a7f2b4fe28b326955dc459e4d949a1) in pr https://github.com/apache/james-project/pull/2358

**Review by Arsnael — APPROVED** [review]

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisConfigurationTest.scala [hunk 1]
```diff
@@ -88,4 +88,32 @@ class RedisConfigurationTest extends AnyFlatSpec with Matchers {
...
+    config.setListDelimiterHandler(new DefaultListDelimiterHandler(','))
+    config.addProperty("redisURL", "redis-sentinel://secret1@redis-sentinel-1:26379,redis-sentinel-2:26379,redis-sentinel-3:26379?sentinelMasterId=mymaster")
+    config.addProperty("redis.topology", "master-replica")
+
+    val redisConfig: RedisConfiguration = RedisConfiguration.from(config)
+    redisConfig.isInstanceOf[MasterReplicaRedisConfiguration] shouldEqual (true)
+    val redisMasterReplicaRedisConfiguration = redisConfig.asInstanceOf[MasterReplicaRedisConfiguration]
+
+    redisMasterReplicaRedisConfiguration.redisURI.value.size shouldEqual 1
+    redisMasterReplicaRedisConfiguration.redisURI.value.head.toString shouldEqual "redis-sentinel://*******@redis-sentinel-1,redis-sentinel-2,redis-sentinel-3?sentinelMasterId=mymaster"
```
Comment (quantranhong1999) [inline backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisConfigurationTest.scala:103]: Q: What are those **** ?

**Review by quantranhong1999 — APPROVED** [review]

File: backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisConfigurationTest.scala (hunk 1 above)
Comment (vttranlina) [reply backends-common/redis/src/test/java/org/apache/james/backends/redis/RedisConfigurationTest.scala:103]: The RedisURI (lettuce library) auto hidden sensitive info

File: backends-common/redis/src/main/java/org/apache/james/backends/redis/RedisConfiguration.scala [hunk 2]
```diff
@@ -104,11 +105,17 @@ case class StandaloneRedisConfiguration(redisURI: RedisURI, ioThreads: Option[In
 }
 
 object MasterReplicaRedisConfiguration {
-  def from(config: Configuration): MasterReplicaRedisConfiguration =
-    from(config.getStringArray("redisURL"),
+  def from(config: Configuration): MasterReplicaRedisConfiguration = {
+    val redisUris: Array[String] = config.getString(REDIS_URL_PROPERTY_NAME) match {
+      case rawValue if rawValue.startsWith("redis-sentinel:") => Array(config.getStringArray(REDIS_URL_PROPERTY_NAME).mkString(","))
```
Comment (hungphan227) [inline backends-common/redis/src/main/java/org/apache/james/backends/redis/RedisConfiguration.scala:110]: `config.getStringArray(REDIS_URL_PROPERTY_NAME).mkString(",")`

Why not simple config.getString(REDIS_URL_PROPERTY_NAME) ?

Comment (vttranlina) [reply backends-common/redis/src/main/java/org/apache/james/backends/redis/RedisConfiguration.scala:110]: I tried, but it didn't work

Comment (hungphan227) [reply backends-common/redis/src/main/java/org/apache/james/backends/redis/RedisConfiguration.scala:110]: ok

**Comment by Arsnael** [comment]
Related? https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2362/1/

**Comment by Arsnael** [comment]
Rebase please

**Review by Arsnael — APPROVED** [review]

**Pull request merged**